Robustness improvements to HDA codec probe

kv2019i opened the pull request:
Here's a set of patches to expand usage of the generic HDA driver and make the SOF HDA stack handling of codec probe errors more robust/flexible.

Fixes: #1558 
Fixes: #1367 (updated, I had wrong bug number here)

The main idea is that the generic HDA driver is the last resort and only tried in the case more specific machine driver is not found. With current code, if more codecs are found than what the generic driver can handle, probe error is returned and user is left with no audio. The patchset loosens the logic such that the generic driver is used to enable a subset of hardware, and only a warning is emitted. If user wants to enable the full hardware capabilities, a more specific machine driver needs to be written/enabled. It is expected these cases to be very rare.

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -171,6 +176,7 @@ static int skl_hda_audio_probe(struct platform_device *pdev)
 	if (!mach)
 		return -EINVAL;
 
+	hda_soc_card.dev = &pdev->dev;
```
is this not a different issue? or is this even required to move this  here?

ranj063 replied:
or for that matter required at all. devm_snd_soc_register_card() might actually handle this already?

kv2019i replied:
> is this not a different issue? or is this even required to move this here?

@plbossart This is a bit out-of-place, but I wanted to use dev_warn() in skl_hda_fill_card_info and easiest way to do this was to assign hda_soc_card.dev early.

plbossart commented on sound/soc/sof/intel/hda-codec.c:
```diff
@@ -154,12 +154,15 @@ int hda_codec_probe_bus(struct snd_sof_dev *sdev,
-			return ret;
+			dev_warn(bus->dev, "codec #%d probe error, ret: %d\n",
+				 i, ret);
+			bus->codec_mask &= ~BIT(i);
```
right, but if you do this there could be a case where the external HDAudio is detected, but the iDISP isn't and the machine driver does not support this case. we also don't have topologies for HDAudio only.

plbossart replied:
shouldn't we also make the codec_mask here so that we only have 2 bits valid here.

kv2019i replied:
The idea here is to just prune codecs that are not responding to commands. We have a couple of bugs where the STATESTS line is up, but the actual codec is not present or is faulty. The outcome may be a codec mask which the machine drivers cannot handle, but then it's ok to fail later at machine_check. Not sure I got what you mean by "making the codec_mask", I do update bus->codec_mask here which is the thing we pass around.

ranj063 replied:
@kv2019i I feel like patch 1 and 3 are not required if you just have patch 2. You'd have adjusted the codec_mask so that the count will always be either 1 or 2 now?

kv2019i replied:
Ack, 1+3 and 2 are for difference scenarios. See longer reply below ( https://github.com/thesofproject/linux/pull/1612#discussion_r356177999 )

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -120,6 +120,11 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+		dev_warn(card->dev, "Only max two codecs supported!\n");
+		codec_count = 2;
+	}
+
```
how would this work? You have 3 codecs, how would you select two out of three.

ranj063 replied:
@kv2019i instead of manipulating the codec count, why not change the else if down below on line 143 to handle the case count >=2 instead of count == 2?

kv2019i replied:
@plbossart Uhm, you have a point. I was assuming we enumerate the codecs in deterministic order, and we select the codec with lowest address bit first, so SOF PCM pipe is set up, the first HDA codec will be included in the pipeline. But, but, going through the code again, I'm not 100% sure the first codec is selected. Hmm, I'll see if I can guarantee this somehow. If not, I can reduce the patchset so that we only ignore probe failures.

kv2019i replied:
Ok, I think this actually does work, I double-checked it now. We hardcode the bus address in machine drivers (the "ehdaudioXDy" names), so it is guaranteed we will set up the right codec -- we always enable addr=0 as the external HDA codec, no matter if other codecs are found.

But, but, I admit this is less important. I'll split it out for now and put the multi-codec case to the enhancement list for now.

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -692,14 +692,10 @@ static int hda_generic_machine_select(struct snd_sof_dev *sdev)
+		 * hda machine driver if at least the HDMI codec is present
 		 */
-		if (!pdata->machine && codec_num <= 2 &&
-		    HDA_IDISP_CODEC(bus->codec_mask)) {
```
@kv2019i will this case ever be true that codec_num will be > 2? In the previous patch, you'd have adjusted the codec_mask to drop the codec that failed to probe isnt it?

kv2019i replied:
@ranj063 There are two different scenarios tackled. First is that we have a faulty external codec. In this case, we may go from iDisp+Ext=>IDisp_only, or from IDisp+ext+ext2=>iDisp+ext1. 

Second case we have a system that has multiple working external HDA codecs attached on the bus. In this case the probes will succeed for all codecs, but we do not have a machine driver (and related infra) to tackle this configuration. What we can do is to still enable at least the first external HDA codec.

But, but, while responding to @plbossart previous comment, I realized guaranteeing we can pick configure the right codec, might be more complex than what I have here. If I can't solve that, then I have to revert to just address the first scenario (i.e. faulty codec probes).

ranj063 replied:
> Second case we have a system that has multiple working external HDA codecs attached on the bus. In this case the probes will succeed for all codecs, but we do not have a machine driver (and related infra) to tackle this configuration. What we can do is to still enable at least the first external HDA codec.

@kv2019i this would be a flawed strategy isnt it? If there is a working codec, we need to provide the right machine driver that manages it?

kv2019i replied:
@ranj063 Well, the generic HDA driver is a fallback in any case. Yes we should have a more specific machine driver, but if we reach this point in code, search for that machine driver already failed. So the question is then that should we at least enable HDMI audio, and maybe a single HDA codec , which we do have machine driver, or leave the end user with no audio at all.

But yeah, not so clear cut. Currently I anyways have pieces missing to tackle this, so I need to do some more homework for this.

ranj063 replied:
@kv2019i with my multi-client stuff, I do plan to separate out the HDMI codec into a separate client. With that change, you'd probably have the fallback with just the HDMI codecs working if they are probed successfully.

kv2019i commented:
@plbossart @ranj063 Dropped patches 1+3 and modified PR to only address the case of a faulty/unresponsive codec. I added the "handle multiple codecs on the bus" as an enhancement to #990  and will return to this later. Patch 2 addresses the two open bugs and should be less controversial change to take in.

plbossart commented on sound/soc/sof/intel/hda-codec.c:
```diff
@@ -154,12 +154,15 @@ int hda_codec_probe_bus(struct snd_sof_dev *sdev,
 
+	if (!bus->codec_mask)
+		return -ENODEV;
+
```
/* create codec instances */
	hda_codec_probe_bus(sdev, hda_codec_use_common_hdmi);

so that -ENODEV would have no impact...

kv2019i replied:
@plbossart That matches the existing implementation, but you are right the errors (if passed) should be handled. I now updated the patch and made the probe function return void. I did also try a version that returns the error and passes it up from the caller, but we actually get more understandable error messages if we let this pass and get an error message from machine_check(). Who knows, maybe there's some case where DSP will be loaded despite no codecs found.
